src/algorithms/preprune.rs

```diff
@@ -0,0 +1,116 @@
+use crate::{Correctness, Guess, Guesser, DICTIONARY};
+use once_cell::sync::OnceCell;
+use std::borrow::Cow;
+
+static INITIAL: OnceCell<Vec<(&'static str, usize)>> = OnceCell::new();
+static PATTERNS: OnceCell<Vec<[Correctness; 5]>> = OnceCell::new();
+
+pub struct PrePrune {
+    remaining: Cow<'static, Vec<(&'static str, usize)>>,
+    patterns: Cow<'static, Vec<[Correctness; 5]>>,
+}
+
+impl PrePrune {
+    pub fn new() -> Self {
+        Self {
+            remaining: Cow::Borrowed(INITIAL.get_or_init(|| {
+                Vec::from_iter(DICTIONARY.lines().map(|line| {
+                    let (word, count) = line
+                        .split_once(' ')
+                        .expect("every line is word + space + frequency");
+                    let count: usize = count.parse().expect("every count is a number");
+                    (word, count)
+                }))
+            })),
+            patterns: Cow::Borrowed(PATTERNS.get_or_init(|| Correctness::patterns().collect())),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct Candidate {
+    word: &'static str,
+    goodness: f64,
+}
+
+impl Guesser for PrePrune {
+    fn guess(&mut self, history: &[Guess]) -> String {
+        if let Some(last) = history.last() {
+            if matches!(self.remaining, Cow::Owned(_)) {
+                self.remaining
+                    .to_mut()
+                    .retain(|(word, _)| last.matches(word));
+            } else {
+                self.remaining = Cow::Owned(
+                    self.remaining
+                        .iter()
+                        .filter(|(word, _)| last.matches(word))
+                        .copied()
+                        .collect(),
+                );
+            }
+
+            // prune patterns
```

Comment:
I think we could actually do this in a slightly nicer way that doesn't require `compatible_pattern`. Here's what I'm thinking (inspired by `enumeration`). First, remove the prune up here, and add a field `matches: [usize; MAX_MASK_ENUM]` to `Candidate`. Then, add

```rust
let totals = [0usize; MAX_MASK_ENUM];
```

to the top of the `for` loop below. Then see my comment there.